spring-transaction/src/test/kotlin/org/jetbrains/exposed/spring/SpringTransactionEntityTest.kt

```diff
@@ -59,36 +68,54 @@ open class Service {
     open fun findOrderByProduct(product: String): OrderDAO? {
         return OrderDAO.find { OrderTable.product eq product }.singleOrNull()
     }
+
+    open fun transaction(block: () -> Unit) {
+        block()
+    }
+
+    open fun cleanUp() {
+        SchemaUtils.drop(CustomerTable, OrderTable)
+    }
 }
 
 open class SpringTransactionEntityTest : SpringTransactionTestBase() {
 
     @Autowired
     lateinit var service: Service
 
-    @Test @Commit
-    open fun test01() {
-        transaction {
-            SchemaUtils.create(CustomerTable, OrderTable)
-        }
+    @Autowired
+    lateinit var ds: EmbeddedDatabase
+
+    @BeforeTest
+    open fun beforeTest() {
+        service.init()
+    }
 
+    @Test
+    @Commit
+    open fun test01() {
         val customer = service.createCustomer("Alice1")
         service.createOrder(customer, "Product1")
         val order = service.findOrderByProduct("Product1")
         assertNotNull(order)
-        transaction {
+        service.transaction {
             assertEquals("Alice1", order.customer.name)
         }
     }
 
-    @Test @Commit
+    @Test
+    @Commit
     fun test02() {
         service.doBoth("Bob", "Product2")
         val order = service.findOrderByProduct("Product2")
         assertNotNull(order)
-        transaction {
+        service.transaction {
             assertEquals("Bob", order.customer.name)
-            SchemaUtils.drop(CustomerTable, OrderTable)
         }
     }
```

Comment:
This is an inherent problem with using transaction { } in Spring, if it's an uninitialized transaction inside a Spring Container, it will fetch the last of the initialized Datasources.

In the case where CI failed, the problem was caused by importing the embeddedTest2 database of the newly added SpringMultiContainerTransactionTest and throwing it away.

So I tried to change the configuration to determine the transaction boundary and use it.

Reply:
This issue has been around since before this PR, and it seems like a very tricky issue to resolve since you're managing the TransactionManager as a companion object.